Port C# hardware intrinsics APIs for SSE from SIMD native algorithms

briancylui opened the pull request:
- with unit tests and performance tests for two target frameworks: .NET Core App 3.0 and .NET Standard 2.0.
- .NET Core App 3.0 gets the new C# hardware intrinsics APIs, while .NET Standard 2.0 gets the original native SIMD algorithms.
- Several things have made this multi-targeting feature possible.
   1. The new CpuMathUtils class that contains the new APIs is implemented as a partial class with method definitions split into two separate files (src\Microsoft.ML.CpuMath\CpuMathUtils.[target].cs), only one of which is compiled based on the target framework.
   2. The Microsoft.ML.CpuMath.csproj file makes the switching decision to compile the right files based on the target framework.

### Structure:
1. All new hardware intrinsics APIs are contained in src\Microsoft.ML.CpuMath.
2. Unit tests for the two target frameworks live in test\Microsoft.ML.CpuMath.UnitTests.[target], and contain the same content except for the target framework in .csproj.
3. Performance tests live in test\Microsoft.ML.CpuMath.PerformanceTests.

### Changes to users:
1. Originally, users call intrinsics APIs via the SseUtils class in src\Microsoft.ML.CpuMath\Sse.cs, but now they call them via the new CpuMathUtils class, which will handle switching between SSE and AVX in the future.
2. CpuMathUtils methods and SseUtils methods share the same signature, but CpuMathUtils methods additionally call a new helper class (SseIntrinsics) for C# implementations of SSE operations.

### Future follow-up for `CpuMath` enhancement
1. Suggestions on `CpuMath` enhancement in this PR scheduled for future follow-ups have been compiled into an issue page [here](https://github.com/briancylui/machinelearning/issues/2).

eerhardt commented:
> <Project Sdk="Microsoft.NET.Sdk">

Everything in this `src\Native\CpuMath` folder should be deleted. We don't need it in the official code base.

---
Refers to: src/Native/CpuMath/Intrinsics.csproj:1 in f0f81a5. [](commit_id = f0f81a5019a3c8cbd795a970e40d633e9e1770c1, deletion_comment = False)

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/UnitTests.cs:
```diff
@@ -0,0 +1,451 @@
+using System;
```
All code files need a copyright header on them.  See other `.cs` files for what the header should look like.

briancylui replied:
Thank you.  Some new files include the header, while some others don't.  Will fix it right away.

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/UnitTests.cs:
```diff
@@ -0,0 +1,451 @@
+
+namespace Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp
+{
+    public class UnitTests
```
I think we should give this class a more descriptive name of what it is testing.  Probably something similar to `CpuMathUtilsTests`.

briancylui replied:
Edited thanks

eerhardt commented on DotnetCLIVersion.txt:
```diff
@@ -1 +1 @@
-2.1.200
\ No newline at end of file
+2.1.301
```
Let's revert this change for now.  We needed it when we were using `netcoreapp2.1`, but now that we are using `netcoreapp3.0`, this change isn't necessary.

briancylui replied:
Edited thanks

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+    {
+        public const int CbAlign = 16;
+
+        private static bool Compat(AlignedArray a)
```
I don't see this nor the `Ptr` method being called.  Can they be removed?

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+{
+    public static partial class CpuMathUtils
+    {
+        public const int CbAlign = 16;
```
I don't think this needs to be public.  Also, `16` aligned is specific to SSE, and when we want to do AVX it would need to be changed.
Regardless, this constant can be removed if the following 2 methods are removed.

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+            Contracts.Assert(0 <= offset && offset < dst.Length - count);
+
+            ArraySegment<float> dstSeg = new ArraySegment<float>(dst, offset, count);
+            Scale(a, dstSeg.Array, count);
```
I'm not sure if this usage of `ArraySegment` is correct.

dstSeg.Array is just going to return the original array `dst`, so we are missing using the offset at all.

We should invert the way we call these two methods instead.  `Scale(float a, float[] dst, int count)` should call `Scale(float a, float[] dst, int offset, int count)` with `offset == 0`.

tannergooding replied:
Updating these, long term, to use `Span<T>` would probably be better as well.

Note: Some of my comments, like this, can be logged as "future" work items. Having the initial PR be a "clean" port of the existing code is desirable. Some amount of refactoring or general improvements can happen in a later PR.

briancylui replied:
Nice catch.  I agree that the usage of `ArraySegment` does not achieve the intended result, and that we should have  `Scale(float a, float[] dst, int offset, int count)` call `Scale(float a, float[] dst, int count)` with `offset == 0`.  

The latter means that we should have the Scale method with the offset call an SSE intrinsic, e.g. ScaleU.  However, currently, there is no implementation of SSE intrinsics that accepts an offset.  The native implementation works well with the offset ([here](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/Sse.cs#L635)) because we can increment the input pointer with the offset in C, but in the managed implementation of SSE intrinsics, we need to pass in an array that starts with the `offset`-th element.  Since the Scale functions intends to modify the input arrays in place, making any copy of the original input arrays may not work.  One suggestion is to create new SSE intrinsics in `SseIntrinsics.cs` that accept an offset, and have the original SSE intrinsics call those new intrinsics with `offset == 0`.  This suggested approach will make `SseIntrinsics.cs` no longer completely in sync with `src\Native\Sse.cpp`, but could solve the problem.

Please let me know how the suggested approach sounds.  If it sounds ok, I will modify in this fashion right away.

tannergooding replied:
The "proper" way to fix this should be to just use `Span<T>` (IMO). `Span<T>` allows you to do exactly as you suggest and treat `array[offset]` as if it was index 0.

Outside of that, it should be trivial to update `ScaleU` to take an offset (should just be changing `float* pSrcCurrent = psrc` to `float* pSrcCurrent = psrc + offset` and the same for pdst`)

briancylui replied:
Thank you so much! Edited :)

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+            }
+            else
+            {
+                for (int i = 0; i < dst.Length; i++)
```
This should do `count` number of elements, and not `dst.Length` number of elements.

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+            }
+            else
+            {
+                // Software fallback.
```
It seems odd to have this comment in just one method.  I think it can be removed as it doesn't provide much value IMO.

eerhardt commented on src/Microsoft.ML.CpuMath/Microsoft.ML.CpuMath.csproj:
```diff
@@ -16,7 +15,16 @@
 
   <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
+    <Compile Remove="CpuMathUtils.DotNetStandard.cs" />
+    <None Include="CpuMathUtils.DotNetStandard.cs" />
```
Maybe add a comment and a link to [https://github.com/dotnet/project-system/issues/935](https://github.com/dotnet/project-system/issues/935) to tell people why we are doing this.

Note: I would have thought just adding an unconditional `<None Include="**/*.cs"/>` would be sufficient.  That way we only need to do this in one place, instead of each specific file.

briancylui replied:
Agree with adding a comment - thanks.

Regarding the note: I thougth we only add a "None Include" line for those files removed from compilation, which are different for the two target frameworks.  Pulling `<None Include="**/*.cs"/>` outside of a conditional item group may require three `<None Include="**/*.cs"/>` , which are not all necessary.  But I'm not sure - please let me know.

briancylui replied:
Any idea?  Thanks!

eerhardt replied:
This works for me:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <Configurations>Debug;Release;Debug-Intrinsics;Release-Intrinsics</Configurations>
    <UseIntrinsics Condition="'$(UseIntrinsics)' == ''">$(Configuration.EndsWith('-Intrinsics'))</UseIntrinsics>
    <TargetFramework Condition="'$(UseIntrinsics)' != 'true'">netstandard2.0</TargetFramework>
    <TargetFrameworks Condition="'$(UseIntrinsics)' == 'true'">netstandard2.0;netcoreapp3.0</TargetFrameworks>
    <IncludeInPackage>Microsoft.ML.CpuMath</IncludeInPackage>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <DefineConstants>$(DefineConstants);CORECLR;PRIVATE_CONTRACTS</DefineConstants>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="..\Microsoft.ML.Core\Utilities\Contracts.cs" />

    <!-- Workaround https://github.com/dotnet/project-system/issues/935 -->
    <None Include="**/*.cs" />
  </ItemGroup>

  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
    <Compile Remove="CpuMathUtils.DotNetStandard.cs" />
    <!-- This is only needed until https://github.com/dotnet/corefx/issues/31064 is addressed. -->
    <PackageReference Include="System.Runtime.Intrinsics.Experimental" Version="4.6.0-preview1-26708-04" />
  </ItemGroup>

  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
    <Compile Remove="CpuMathUtils.DotNetCoreApp.cs" />
    <Compile Remove="SseIntrinsics.cs" />
  </ItemGroup>
</Project>
```

eerhardt commented on src/Microsoft.ML.CpuMath/Microsoft.ML.CpuMath.csproj:
```diff
@@ -16,7 +15,16 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
+    <Compile Remove="CpuMathUtils.DotNetStandard.cs" />
```
Let's use the normal TFM naming and casing on this file suffix, that's how we do it in corefx:

`CpuMathUtils.netstandard.cs`
`CpuMathUtils.netcoreapp.cs`

tannergooding commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+        public static void Scale(float a, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count && count <= dst.Length);
```
nit: I think that `count > 0` is easier to read than `0 < count`

tannergooding commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count);
+            Contracts.Assert(0 <= offset && offset < dst.Length - count);
```
nit: `offset < dst.Length - count` is more readable as `offset < (dst.Length - count)`; otherwise, you have to stop and think about operator precedence

briancylui replied:
Agree.  Due to conflicts with the exisiting convention, saving this change for a future follow-up.

tannergooding commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+            Contracts.Assert(0 < count && count <= dst.Length - dstOffset);
+
+            ArraySegment<float> dstSeg = new ArraySegment<float>(dst, dstOffset, count);
+            AddScale(a, src, dstSeg.Array, count);
```
As @eerhardt pointed out above, passing `dstSeg.Array` is not correct, as it just returns the underlying array and doesn't start from `dstOffset`: http://source.dot.net/#System.Private.CoreLib/shared/System/ArraySegment.cs,62

briancylui replied:
Very much agree.  Thanks for pointing it out.  Reference: https://github.com/dotnet/machinelearning/pull/562/files/f0f81a5019a3c8cbd795a970e40d633e9e1770c1#diff-002eee6f7c1e6031c143a2fb39cfe57cR54.

eerhardt commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
```
@tannergooding - do you think it is worthwhile to do the "preamble", like we do in https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs#L96-L102?

@briancylui - this comment wouldn't need to be addressed in the first PR, but if necessary, we could address it in a future revision.

tannergooding replied:
I think it would be useful to link to the software implementation with a comment indicating that this a vectorized version of the software/serialized version.

eerhardt replied:
What I was specifically asking about is getting aligned at the beginning of the array - that's the "preamble" part I was referring to.

1. while (!aligned) { do scalar operation; }  // preamble
2. Do vectorized operation using Read**Aligned**
3. while (!end) { do scalar operation; }

tannergooding replied:
Ah, I see. Yes, we should. For large arrays, especially those that cross cache line or page boundaries, doing this should save some measurable amount of time.

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetStandard/UnitTests.cs:
```diff
@@ -0,0 +1,450 @@
+using System;
```
Instead of duplicating this file in the 2 different folders, we should be linking both .csproj files to the same physical .cs file.  That way we only need to maintain a single file.

eerhardt replied:
This is how you do something like that.  The file can live in one of the folders, and the other `.csproj` can do something like:

`<Compile Include="..\Microsoft.ML.Core\Utilities\Contracts.cs" />`

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x39);
```
A comment on what `0x39` means would be nice (XYZW -> YZWX, it looks like)

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x93);
```
Same here (XYZW -> WXYZ, it looks like)

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        // Warning: this operation changes the value of x => do not reuse x
```
This comment isn't relevant, `x` is a value type and any modification in this method won't propagate out.

briancylui replied:
Thanks for the comment.  Since this Store4 function has the output type `void`, the input argument `x` should be changed, or the Store4 function would be a no-op.  I copied this comment from `src\Native\Sse.cpp` (https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L44).

tannergooding replied:
The difference between the native code and the C# code is that the native code is a `#define`, which means that it will directly modify the variable `x` in the context of the containing method.

The C# code is an actual function, and has all the normal safety rules (`x` is a separate local from any `x` in the caller and it is a value type, so any modifications to `x` in `Store4` won't propagate up).

briancylui replied:
Thank you, Tanner!  Since the intended effect of `Store4` is to modify `x`, I will have to change the function a bit.  Should I make `Store4` return `Vector128<float>`, or return `void` with the `out Vector128<float> x` or `ref Vector128<float> x`?  Thank you for your advice in advance!

tannergooding replied:
I don't believe the intended effect was to modify `x` and, from a quick scan, I didn't see any code that relied on `x` being modified.

I would assume the comment was added because people were initially calling `Store4`, reusing `x`, and have failures that took some time to debug.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+        // Warning: this operation changes the value of x => do not reuse x
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
```
This method should just be a single `Sse.Store`, which will store all four elements at once.

tannergooding replied:
Nevermind, just realized that this is for storing four non-linear indices.

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/UnitTests.cs:
```diff
@@ -0,0 +1,451 @@
+                output.WriteLine($"{nativeOutput} == {managedOutput}");
+
+                // assert
+                Assert.True(nativeOutput == managedOutput);
```
I think this was a good approach for initially testing that the new methods produce the same output as the native methods.  However, I don't think this is a good long term approach. In the future, we some day hope to remove the native methods, if possible.  Also, as new methods get written to both the native and C# intrinsics, it would be quite possible that someone would write the same bug in both places.

Instead, I think the test should hard-code what it expects the results to be.

briancylui replied:
Agree - that would save us from PInvoke calls in unit tests too.  Will fix that - thanks!

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/UnitTests.cs:
```diff
@@ -0,0 +1,451 @@
+                output.WriteLine($"{nativeOutput} == {managedOutput}");
+
+                // assert
+                Assert.True(nativeOutput == managedOutput);
```
In general, don't use `Assert.True(a == b)`.  Instead use `Assert.Equal(a, b)`.  This way when the test fails, it fails and says "Actual: 3. Expected 4" instead of just a general "Failed" message.

briancylui replied:
Agree and will fix that - thanks!

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
```
I think we could do this more efficiently. Generally you would do something like: `var remainder = count % elementsPerIteration;` and then `float* pEnd = pdst + (count - remainder);`. Your loop check then just does `while (pDstCurrent < pEnd)`

tannergooding replied:
This is probably good for future followup.

Generally speaking, a clean 1-to-1 port with later refactoring is preferred to a port + refactoring in one. This makes it, all together, easier to review, track bugs, etc.

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs:
```diff
@@ -0,0 +1,362 @@
+            return q;
+        }
+
+        public static void Scale(float a, float[] dst, int count)
```
I agree they are unclear.

But I don't think we should we should do a one-off change here and rename these variables, but the existing, corresponding variables in Sse.cs be left as `a`, `dst`.  If we intend on renaming, I think we should do a mass renaming across the board for all these functions.

tannergooding replied:
Right, I think this falls under my general comment of: A clean port at first is desirable, so a number of my comments can be read as "we should do this later in a follow up"

briancylui replied:
Yes, I agree that they are unclear too.  Originally I named the first argument "scale" instead of "a", but the existing code used "a" so I followed the existing convention.  Thanks for raising the naming issue.

tannergooding replied:
> If we intend on renaming, I think we should do a mass renaming across the board for all these functions.

Yes. This should probably be done as a cleanup item.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                        pDstCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
```
Generally speaking, you can finish off the remaining indices one of two ways.

Like this, where you drop down to a scalar algorithm or, you can "double compute" a couple of the indices, that involves:
* Saving the stored result (`dstVector`) from the last iteration of the vectorized code
* Moving `pDstCurrent` back such that `pDstCurrent + elementsPerIteration == pEnd`
* Doing a single iteration for the remaining elements
* Mix the saved result from the last iteration of the vectorized code with the result from the remaining elements
* Write the result

This generally results in more performant code, depending on the exact algorithm and number of remaining elements

briancylui replied:
Thank you!  This is much needed.  Same as above: Should I save it for a future follow-up, or should I propagate this fix to other functions at this point?

tannergooding replied:
This is probably good for future followup.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+            }
+        }
+
+        internal static void AddScaleU(float scale, float[] src, float[] dst, int count)
```
Does the `U` here (and elsewhere) mean "Unaligned" or something else?

briancylui replied:
Yes, you're right - `U` means unaligned. Reference: https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L5-L20

Also adding a comment at the top of `SseIntrinsics.cs` for clarity.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
```
Made a comment above (for `ScaleU`) it is also applicable here, and below on similar code.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        srcVector = Sse.Multiply(srcVector, scaleVector);
```
On newer hardware, this can become a `FusedMultiplyAdd`, provided a different (but more precise) result is acceptable for some inputs.

tannergooding replied:
It would be part of any AVX related code-work.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
+            for (int i = 1; i <= 3; i++)
```
Unrolling this would likely be more performant.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+            }
+        }
+
+        internal static void AddScaleSU(float scale, float[] src, int[] idx, float[] dst, int count)
```
Does `S` here stand for `Sparse`, or something else?

eerhardt replied:
Yes, and `U` stands for unaligned.  See https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L5-L15.

We don't necessarily need to follow the same naming pattern here.  It may make sense to keep them aligned.  But it may be more confusing if/when the native assembly is removed.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+
+                    while (pIdxCurrent < pEnd)
+                    {
+                        pDstCurrent[*pIdxCurrent] += scale * *pSrcCurrent;
```
nit: please add parenthesis where it improves readability.

You are mixing multiplication and pointers here, so `scale * *pSrcCurrent` makes you have to stop and manually process the operands apart.

briancylui replied:
Very much agree.  Thank you!

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
```
The second `Sse.MoveHighToLow(result, Sse.UnpackLow(result, result))` is probably more efficient as: `Shuffle(result, result, (1, 0, 3, 2))`

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+            }
+            else
+            {
+                mask = Sse.SetAllVector128(0x7FFFFFFF);
```
This isn't right. This will result in `Sse.SetAllVector128((float)(0x7FFFFFFF))`

tannergooding replied:
You probably want: `Sse.SetAllVector128(BitConverter.Int32BitsToSingle(0x7FFFFFFF))`

briancylui replied:
Edited thanks!  Is the line above (Line 324) still correct?

briancylui replied:
Any idea?  Thanks!

eerhardt replied:
Yes, line 324 is correct since there is an overload in `Sse2` for `public static Vector128<int> SetAllVector128(int value);`.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                    }
+
+                    // TODO: Check code length per line
+                    if (Sse3.IsSupported)
```
Abstracting this out into a helper `VectorSum` method might be useful (it is used here and above)

briancylui replied:
Thanks - fixed with a helper `VectorSum` method with return type `Vector128<float>`.  Do you think it's better for `VectorSum` to change the input array in place and have the return type be `void` instead?

tannergooding replied:
No, returning `Vector128<float>` is better here as the existing code you pulled out into the helper method doesn't modify the array.

tannergooding commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> distanceVector = Sse.Subtract(Sse.LoadVector128(pSrcCurrent), Sse.LoadVector128(pDstCurrent));
+                        SqDistanceVector = Sse.Add(SqDistanceVector, Sse.Multiply(distanceVector, distanceVector));
```
I think there might have been other cases above as well, by why aren't you just doing `SqDistanceVector = Sse.Multiply(distanceVector, distanceVector)`

tannergooding commented on src/Native/CpuMath/IntrinsicsUtils.cs:
```diff
@@ -0,0 +1,796 @@
+using System;
+
+namespace Intrinsics
+{    public class IntrinsicsUtils
```
Many of these methods look duplicated from `SseIntrinsics.cs`?

briancylui replied:
Yes, nice catch.  Sorry shouldn't have incorporate this `src\Native\CpuMath` folder in this PR.  Sorry for any inconvenience caused, and thank you for your understanding.

tannergooding commented on src/Native/CpuMath/IntrinsicsUtils.cs:
```diff
@@ -0,0 +1,796 @@
+        }
+
+        //// Multiply matrix times vector into vector.
+        //private static float MatMulASse(bool add, float[] matrix, float[] src, float[] dst)
```
If these aren't complete, ready for use yet, we shouldn't include them as commented out as part of this PR.

danmosemsft replied:
Right, normally please do'nt checkin commented code.

danmosemsft commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
```
same for the others I think.

danmosemsft commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
```
I don't think this comment adds much (it literally restates the statement immediately proceeding) so I would remove these.

`// SEE3 is not supported` is OK because it saves you having to read upwards to find the conditional clauses.

danmosemsft commented on src/Native/CpuMath/IntrinsicsUtils.cs:
```diff
@@ -0,0 +1,796 @@
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
```
extra newline -- just replace `^\s*\n\s*\n` with `\n` throughout.

danmosemsft commented on src/Native/CpuMath/IntrinsicsUtils.cs:
```diff
@@ -0,0 +1,796 @@
+            }
+            else
+            {
+                // Software fallback.
```
Use consistent text throughout eg `// SSE is not supported`. Also avoid periods unless it's a sentence.

danmosemsft commented on src/Native/CpuMath/IntrinsicsUtils.cs:
```diff
@@ -0,0 +1,796 @@
+
+            unsafe
+            {
+
```
Another extra newline. I will stop commenting on these - just search and replace.

danmosemsft commented on src/Native/CpuMath/IntrinsicsUtils.cs:
```diff
@@ -0,0 +1,796 @@
+                        pDstCurrent += 4;
+                    }
+
+                    // TODO: Check code length per line
```
What does this mean - the characters on the line? I don't think we have a firm rule. Use judgement. No need to check in this comment.

danmosemsft replied:
Perhaps it means something else though?

briancylui replied:
Thanks!  Good to know - I thought I had to make every line less than 79 characters long for the sake of the reader as in those systems assignment.  Just now fixed the code length per line anyway :)  Thank you!

Sorry I should have omitted this 'src\Native\CpuMath` folder from the PR - it was my initial working repo with commented lines, unfixed styles, and TODOs for my development.  Please feel free to ignore this folder, and the codes in `Microsoft.ML.*` are those that get finally integrated into the dotnet repo.  Sorry for any inconvenience caused, and thank you for your understanding.

briancylui commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp.csproj:
```diff
@@ -0,0 +1,18 @@
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.ML" Version="0.2.0" IncludeAssets="Native" />
+    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
```
Removing this line works fine on local machine.  Edited thanks.

briancylui commented on test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp.csproj:
```diff
@@ -0,0 +1,18 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.ML" Version="0.2.0" IncludeAssets="Native" />
```
Removing this line causes error in local machine: ` System.DllNotFoundException : Unable to load DLL 'CpuMathNative' or one of its dependencies: The specified module could not be found. (Exception from HRESULT: 0x8007007E)`.  Does removing this line work elsewhere in the preview package?

eerhardt replied:
The way the unit tests get the 'CpuMathNative' assembly copied is through this mechanism:

https://github.com/dotnet/machinelearning/blob/31ae67897e0cec6b9a3fa25c4a23dbec23533415/test/Microsoft.ML.Predictor.Tests/Microsoft.ML.Predictor.Tests.csproj#L28

briancylui commented on src/Native/CpuMath/tests/Performance/Perf.cs:
```diff
@@ -0,0 +1,245 @@
+            src1 = new float[len];
+            src2 = new float[len];
+            original = new float[len];
+            Random rand = new Random(seed);
```
This is such a good suggestion.  In our case, we want to generate and reproduce an input array of many elements, so a seed is introduced to set the entire array.  It will be useful to the tester to be able to test functions on different input arrays by setting different seeds.  

Not sure about the preferred style of writing performance tests.  Do we usually:
(1) write a hard-coded perf test with provided seeds that ensure everyone who runs the code to get the same, standardized result, or
(2) write a perf test that allows testers to customize the seeds and hence test with different input arrays?
Please let me know - thanks!

tannergooding replied:
I believe the general preference is to hard code the seed, but providing a mechanism to override the seed for additional (local/stress) testing is generally useful.

eerhardt commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,536 @@
+
+                    while (pSrcCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
```
I know that this is ported directly from the C++ code. And while I can't exactly state what kind of issues it will cause, it seems wrong/unsafe to be reading memory outside of the bounds the method is given. We don't have any idea what is after the array in memory, but we are reading/loading at least 12 bytes of memory after the array into this Vector128.

Would it make more sense to turn `result` into a float before this line, and then do scalar abs/add at the end of the array?

tannergooding replied:
This could cause AV, or any number of other issues.

We should either drop back to scalar or do a double-compute for the last chunk. See https://github.com/dotnet/machinelearning/pull/562/files/f0f81a5019a3c8cbd795a970e40d633e9e1770c1#r204073474 for more details.

briancylui commented:
Re-opened this PR to refresh the modified build definition.  Included responses to PR comments.  Since I've deleted a deprecated `src\Native\CpuMath\` folder and renamed some files, some previous PR comments may appear as outdated.  Previous suggestions on enhancing CpuMath are compiled into a list in this issue page: https://github.com/briancylui/machinelearning/issues/2.

This PR is now ready for a second review.  Thank you.

tannergooding commented on Microsoft.ML.sln:
```diff
@@ -112,10 +118,10 @@ Global
-		{46F2F967-C23F-4076-858D-33F7DA9BD2DA}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{46F2F967-C23F-4076-858D-33F7DA9BD2DA}.Debug-Intrinsics|Any CPU.ActiveCfg = Debug-Intrinsics|Any CPU
-		{46F2F967-C23F-4076-858D-33F7DA9BD2DA}.Debug-Intrinsics|Any CPU.Build.0 = Debug-Intrinsics|Any CPU
+		{46F2F967-C23F-4076-858D-33F7DA9BD2DA}.Debug|Any CPU.ActiveCfg = Debug-Intrinsics|Any CPU
```
This looks wrong....

briancylui commented on src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs:
```diff
@@ -0,0 +1,397 @@
+                return norm;
+            }
+        }
+
```
Extra line - will be fixed.

briancylui commented on test/Microsoft.ML.CpuMath.PerformanceTests/SsePerformanceTests.cs:
```diff
@@ -0,0 +1,238 @@
+        }
+
+        [Benchmark]
+        public float MyDotUPerf() => CpuMathUtils.DotProductDense(src, dst, LEN);
```
Will replace all "My" prefixes to "Managed"

eerhardt commented on src/Microsoft.ML.CpuMath/Microsoft.ML.CpuMath.csproj:
```diff
@@ -3,20 +3,30 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <DefineConstants>$(DefineConstants);CORECLR;PRIVATE_CONTRACTS</DefineConstants>
+    <!-- The SDK we currently restore and use has 2.0 as the maximum supported version -->
+    <NETCoreAppMaximumVersion>3.0</NETCoreAppMaximumVersion>
```
You shouldn't need this (either here or in the tests).  Instead, anyone who wants to use the `netcoreapp3.0` version needs to install the 3.0 SDK on their machine.

briancylui replied:
This was added in the late afternoon today in an attempt to pass the `VSTS: public-CI` build test below, but was not very successful.

eerhardt replied:
Let's chat this morning about it. There may be some tricks we need to do.

briancylui replied:
Removing the `NETCoreAppMaximumVersion` tag from the current green build results in errors in running `build.cmd -Release` from the root of the ML repo.  How should I proceed?

eerhardt commented on test/Microsoft.ML.CpuMath.PerformanceTests/Program.cs:
```diff
@@ -0,0 +1,29 @@
+                .Run(null, CreateClrVsCoreConfig());
+        }
+
+        private static IConfig CreateClrVsCoreConfig()
```
Why is this needed?  Why can't we just use the default `static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);`?

briancylui replied:
When I removed the custom config function, the program tried to look for a `BenchmarkDotNet.Autogenerated` folder which did not exist, while having a custom config function has solved the error.  Will try the default `.Run(args);` tomorrow.

briancylui replied:
Thanks for the suggestion!  I have tried `.Run(args)`, and it runs well, but if the user does not type any argument, the program returns swiftly and prints something like `Global Running Time: 0.00000010` on the standard output.  I originally wrote the function `CreateClrVsCoreConfig()` in order to comply with the existing code in `Program.cs` of `test\Microsoft.ML.Benchmark`:

https://github.com/dotnet/machinelearning/blob/8087a20f3294671b522d265ccfa07543c5ff6fa5/test/Microsoft.ML.Benchmarks/Program.cs#L29-L44

Please let me know how I should proceed - Thanks!

eerhardt replied:
No worries. We can leave it like this for now.

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.netstandard/UnitTests.cs:
```diff
@@ -0,0 +1,276 @@
+
+            var managedOutput = CpuMathUtils.DotProductDense(src, dst, dst.Length);
+
+            if (test == 0)
```
Typically, this is handled by passing the "expected" value in as InlineData as well:

```C#
[InlineData(0, 13306.0376f)]
[InlineData(1, (13291.9235f)]
public unsafe void DotUTest(int test, float expected)
```

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.netstandard/UnitTests.cs:
```diff
@@ -0,0 +1,276 @@
+
+            for (int i = 0; i < dst.Length; i++)
+            {
+                dst[i] += 1;
```
Why are we doing this?  (repeated else where)

briancylui replied:
Wanted to test the correctness of the dot product of two unequal float arrays.  Without `dst[i] += 1`, `src` and `dst` are the same, and the dot product result may look like that from the `SumSqU` function.  But I agree that it is not entirely necessary.

eerhardt replied:
I think that is a fine reason. It would be good to stick a code comment in here stating that. Then others reading the code know why this is here, and don't delete it.

eerhardt commented on test/Microsoft.ML.CpuMath.PerformanceTests/Microsoft.ML.CpuMath.PerformanceTests.csproj:
```diff
@@ -0,0 +1,29 @@
+  </PropertyGroup>
+  
+  <ItemGroup>
+    <Compile Remove="BenchmarkDotNet.Artifacts\**" />
```
I wonder if there is a property we can set to put these artifacts into the top-level `bin` folder instead.

@tannergooding - do you know?  I'll ping Adam, if not.

tannergooding replied:
I know you can, by setting `IConfig.ArtifactsPath` (programmatically), but I don't remember how to do that from MSBuild.

eerhardt replied:
I think you can leave it for now.  We have the same in the Benchmark tests that are already checked in:  https://github.com/dotnet/machinelearning/blob/697ea3d73d1dd68d516f46b45124ccd6ebe063f2/test/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj#L8-L12

eerhardt commented on test/Microsoft.ML.CpuMath.PerformanceTests/Microsoft.ML.CpuMath.PerformanceTests.csproj:
```diff
@@ -0,0 +1,29 @@
+    <OutputType>Exe</OutputType>
+    <LangVersion>7.2</LangVersion>
+    <StartupObject>Microsoft.ML.CpuMath.PerformanceTests.Program</StartupObject>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
```
Let's chat about this tomorrow.  This isn't going to work in CI, nor when normal developers try to build.  The reason is because they won't have the 3.0 SDK installed.  We will basically need to skip this project and the UnitTests.netcoreapp project.

eerhardt commented on test/Microsoft.ML.CpuMath.PerformanceTests/Microsoft.ML.CpuMath.PerformanceTests.csproj:
```diff
@@ -0,0 +1,29 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <LangVersion>7.2</LangVersion>
+    <StartupObject>Microsoft.ML.CpuMath.PerformanceTests.Program</StartupObject>
```
Why do you need to set `StartupObject`?

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/Microsoft.ML.CpuMath.UnitTests.netcoreapp.csproj:
```diff
@@ -0,0 +1,19 @@
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Do you still need unsafe code in these unit test projects?

eerhardt commented on src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs:
```diff
@@ -0,0 +1,396 @@
+                for (int i = 0; i < indices.Length; i++)
+                {
+                    int index = indices[i];
+                    dst[index] += a * src[i];
```
Is `src` always considered "dense" here?  and only `dst` is sparse?

briancylui replied:
Yes, you're right.  It seems to me that `src` is always dense, while we only deal with the indexed elements of `dst`. 
 I looked up the native code for this software fallback: https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L1684-L1700.

eerhardt commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,469 @@
+        }
+
+        internal static unsafe void AddScaleU(float scale, Span<float> src, Span<float> dst)
+        {
```
(minor) it may make sense to add some `Debug.Asserts` to check the src and dst Lengths match.  However, these are internal functions that are only called from functions that guarantee the arguments are checked, so I don't believe it is a blocking issue. It just may be some nice documentation on the expectations of these methods.  And in case they get new callsites in the future.

eerhardt commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,469 @@
+
+                while (pDstCurrent < pEnd)
+                {
+                    Vector128<float> src1Vector = Sse.LoadScalarVector128(pSrc1Current);
```
Why do any vector operations here at all?  Why not just simply do the multiplication like normal?

`*pDstCurrent = (*pSrc1Current) * (*pSrc2Current)`

eerhardt replied:
Oh, is it because that's what the existing native C++ code is doing?

---
In reply to: [206389105](https://github.com/dotnet/machinelearning/pull/562#discussion_r206389105) [](ancestors = 206389105)

briancylui replied:
Unfortunately yes :( Link: https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L1830-L1849

eerhardt replied:
OK, that's fine.

tannergooding commented on src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs:
```diff
@@ -0,0 +1,396 @@
+        public static void Scale(float a, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count && count <= dst.Length);
```
For future cleanup: `Span<T>` does most of these checks for you.

tannergooding commented on src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs:
```diff
@@ -0,0 +1,396 @@
+            Scale(a, new Span<float>(dst, 0, count));
+        }
+
+        public static void Scale(float a, float[] dst, int offset, int count)
```
Also for future cleanup: It's probably worth just exposing a public `Span<T>` overload, and letting the external code set offset/count as needed.

eerhardt replied:
That would involve changing the other `partial` class as well, which is used for `netstandard2.0`. It would also mean modifying all the call sites. For sure, we can do that in the future.

tannergooding commented on test/Microsoft.ML.CpuMath.PerformanceTests/SsePerformanceTests.cs:
```diff
@@ -0,0 +1,238 @@
+        {
+            int seed = DEFAULT_SEED;
+
+            if (Environment.GetEnvironmentVariable("CORECLR_SEED") != null)
```
nit: Possibly use something other than `CORECLR_SEED` (maybe `MLNET_SEED`?).

briancylui commented:
Issues at the end of today:
1. Removing `NETCoreAppMaximumVersion` tag from green build results in errors in running `build.cmd -Release`.  Follow-ups needed.
2. Building and running perf tests in local machines fail because:
`The application to execute does not exist: 'C:\Users\t-brlui\repos\machinelearning\bin\AnyCPU.Release\Microsoft.ML.CpuMath.PerformanceTests\netcoreapp3.0\Microsoft.ML.CpuMath.PerformanceTests.dll'`.  Follow-ups needed.

eerhardt commented on build/AfterCommonTargets.targets:
```diff
@@ -0,0 +1,14 @@
+<Project>
+  <PropertyGroup> 
+    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+    <VSTestTargets Condition="'$(UseIntrinsics)' != 'true' and '$(TargetFramework)' == 'netcoreapp3.0'">ignore.targets</VSTestTargets>
```
What does this do?  I'm not familiar.  Also, since you are also overriding `VSTest` - isn't that enough?

eerhardt replied:
If we keep this property setting, we should factor out `'$(UseIntrinsics)' != 'true' and '$(TargetFramework)' == 'netcoreapp3.0'` into an MSBuild property so it is only declared once.

---
In reply to: [206735143](https://github.com/dotnet/machinelearning/pull/562#discussion_r206735143) [](ancestors = 206735143)

briancylui replied:
It turns out that the place where we try to blank `VSTest` (https://github.com/dotnet/machinelearning/pull/562/files/c6f5f8569ea31bc30e0b7ae54870b9cbf51e2bdb#diff-77ebf330684fd059d7aecec3ce7e6218R19) is processed before the program builds the actual `VSTest`, so without this line, even though we would have blanked `VSTest`, the entire `VSTest` would be built all over again.  Because of this order of the build, one way to solve this issue is to ignore the built targets of `VSTest` to ignore the test projects.  Not sure if it is the best way to do it so feel free to suggest any optimization.

safern replied:
I helped him with this yesterday and by doing a pp log I found out that `Microsoft.TestPlatform.targets` where imported after the actual `Empty.targets` file even though they are specified to be imported after the common targets, I don't know if that is the way it is designed or if it is a bug. 

So I found out that there was a `VSTestTargets` overridable property declared in: `15.0\Microsoft.Common.targets\ImportAfter\Microsoft.TestPlatform.ImportAfter.targets:17` that points to `Microsoft.TestPlatform.targets` and this `ImportAfter.targets` file will import that project if the specified path exists. So in order to override the `VSTest` target I suggested to just set that property to a non-existent targets file.

eerhardt replied:
So if I'm reading this correctly, that means this line in Empty.targets:

``` <Target Name="VSTest"/>```

Isn't doing anything and can be removed, right?

Also - it would probably make sense to move setting this property into `Empty.targets`.  That would resolve my "we should factor this condition out" comment above as well.  `build/AfterCommonTargets.targets` is responsible for deciding whether Empty.targets should be imported. Empty.targets decides how to ensure no targets are run (which includes setting this property).

safern replied:
>So if I'm reading this correctly, that means this line in Empty.targets:
>
>`<Target Name="VSTest"/>`
>
>Isn't doing anything and can be removed, right?

It is actually because in run-tests.proj we're calling the VSTest target on every test project: https://github.com/dotnet/machinelearning/blob/master/test/run-tests.proj#L10

And if we don't declare it, it will complain saying that the VSTest target doesn't exist in both test projects targeting netcoreapp3.0.

> Also - it would probably make sense to move setting this property into Empty.targets. That would resolve my "we should factor this condition out" comment above as well. build/AfterCommonTargets.targets is responsible for deciding whether Empty.targets should be imported. Empty.targets decides how to ensure no targets are run (which includes setting this property).

That is right, no sense having the same condition twice, just declare it when `Empty.targets` is imported, which is conditioned already. 

I also thought of just simply excluding both projects from: https://github.com/dotnet/machinelearning/blob/master/test/run-tests.proj#L5

But it makes more sense to me to do everything inside `Empty.targets` since whenever the repo is ready to build and target netcoreapp3.0 we can just remove this files and we're good to go.

eerhardt commented on test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/Microsoft.ML.CpuMath.UnitTests.netcoreapp.csproj:
```diff
@@ -0,0 +1,18 @@
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <IsPackable>false</IsPackable>
+    <!-- The SDK we currently restore and use has 2.0 as the maximum supported version -->
+    <NETCoreAppMaximumVersion>3.0</NETCoreAppMaximumVersion>
```
This should be able to be removed.

eerhardt commented on test/Microsoft.ML.CpuMath.PerformanceTests/Microsoft.ML.CpuMath.PerformanceTests.csproj:
```diff
@@ -0,0 +1,29 @@
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <!-- The SDK we currently restore and use has 2.0 as the maximum supported version -->
+    <NETCoreAppMaximumVersion>3.0</NETCoreAppMaximumVersion>
```
This can be removed.

eerhardt commented:
Let's not merge this until after we branch for `v0.4`.  We shouldn't add risk to `v0.4` with this change...

briancylui commented on src/Microsoft.ML.CpuMath/SseIntrinsics.cs:
```diff
@@ -0,0 +1,471 @@
+                vector = Sse3.HorizontalAdd(vector, vector);
+                vector = Sse3.HorizontalAdd(vector, vector);
+            }
+            else
```
@tannergooding: (Responding to your previous PR comment) Factoring the `VectorSum` function out of the intrinsics has worked quite well, but my old implementation of `VectorSum` (which outputs a `Vector128<float>`) seemed to cause some perf issues (doubled the running time of 4 intrinsics that call `VectorSum`).  I changed it to `ToVectorSum`, which changes `vector` in place, but the same perf issue persisted.  Then, I (thanks to Santi's help) tried the `AggressiveInlining` attribute above to make the function inline, which has improved the perf results back to normal.  Curious about how it sounds to you, and if there is any better way to do it.

Side question: Should I also make the other private functions - `Load1`, `Load4`, `Rotate`, `RotateReverse`, and `Store4` - inline with the `AggressiveInlining` attribute to improve perf?

tannergooding replied:
We'll probably want to change `private static void ToVectorSum(ref Vector128<float> vector)` to `private static Vector128<T> ToVectorSum(in Vector128<float> vector)`, as `Vector128<T>` was made into a `readonly struct`, but it otherwise looks good.

> Should I also make the other private functions - Load1, Load4, Rotate, RotateReverse, and Store4 - inline with the AggressiveInlining attribute to improve perf

I would think they are already being inlined (except for Store4), if that isn't the case, then the methods are probably small enough that inlining would help. I would recommend benchmarking and looking at the codegen for both having them explicitly inlined and not.

helloguo replied:
It does not hurt to put AggressiveInlining attribute on these short and internal functions. 

For your old code, even though VectorSum is inlined, the codegen is not optimized due to register spill and reload. It seems JIT has optimization opportunity over there. Do you mind opening an issue to discuss about it on CoreCLR github repo?

briancylui replied:
Thanks for the input!  Adding `AggressiveInlining` attributes for those short functions to mimic the behavior of the existing native code.

tannergooding replied:
I'll pick up logging a bug for CoreCLR. We should get a minimal repro and ensure it isn't already tracked by one of the "first class struct" issues.

eerhardt commented on src/Microsoft.ML.CpuMath/Microsoft.ML.CpuMath.csproj:
```diff
@@ -2,21 +2,31 @@
     <DefineConstants>$(DefineConstants);CORECLR;PRIVATE_CONTRACTS</DefineConstants>
+    <LangVersion>7.3</LangVersion>
+    <!-- The SDK we currently restore and use has 2.0 as the maximum supported version -->
+    <NETCoreAppMaximumVersion>3.0</NETCoreAppMaximumVersion>
```
This can be removed.

briancylui replied:
Agree with removing the `NETCoreAppMaximumVersion` tag.  

In a recent fix of the `VectorSum` function in `SseIntrinsics.cs`, since we need to rely on the read-only feature of `Vector128` introduced in C# 7.2 to change the vector in-place, the Lang Version has to be at least 7.2.  However, since we use the feature of `extensible fixed statements` in `SseIntrinsics.cs`, which is a feature in C# 7.3, the `LangVersion` restriction has been set to `7.3`.  Reference: https://github.com/dotnet/roslyn/pull/24469

Please let me know how I should proceed.

eerhardt replied:
sorry, I was specifically talking about `<NETCoreAppMaximumVersion>3.0</NETCoreAppMaximumVersion>`.  The `LangVersion` change is needed.

briancylui commented:
Responded to the second round of review, and now this PR is ready for a third round if any.  Gathering thoughts on this PR before fixing the merge conflicts in the `sln` file.  Please let me know if there is anywhere else I should improve my code.  How does it sound to everyone?

briancylui commented:
Merge conflicts have been fixed.  Green builds for all 5 checks.

safern commented on test/Microsoft.ML.CpuMath.PerformanceTests/SsePerformanceTests.cs:
```diff
@@ -0,0 +1,238 @@
+        private const int DEFAULT_SEED = 253421;
+        private const float DEFAULT_SCALE = 1.11f;
+
+        private float[] src, dst, original, src1, src2;
```
Nit: This all non constant privates should be prefixed with `_` but it can be updated in your next PR to not block this one.

briancylui replied:
Updated the issue page (https://github.com/briancylui/machinelearning/issues/2) to include this action item for future follow-up.  Thank you for your input!

briancylui commented:
Thank you all reviewers for your time reviewing this PR and making great suggestions to fix issues or improve performance.  I have learnt a lot from all of you.  

Since I don't have write access to the `dotnet/machinelearning` repo, may anyone with write access please help merging this PR to the master?  Much appreciated!

After the merging, I can then start writing more SSE intrinsics in a new branch.
